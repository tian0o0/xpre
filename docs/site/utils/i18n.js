// component mixin
import { get } from '../../../src/utils';
import { camelize } from '../../../src/utils/format/string';

export default {
  computed: {
    $t() {
      const { name } = this.$options;
      const prefix = name ? camelize(name) + '.' : '';
      const messages = this.$vantMessages[this.$vantLang];

      return (path, ...args) => {
        const message = get(messages, prefix + path) || get(messages, path);
        return typeof message === 'function' ? message(...args) : message;
      };
    }
  }
};
