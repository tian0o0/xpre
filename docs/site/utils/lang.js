import Locale from '../../../src/components/locale';
import zhCN from '../../../src/components/locale/lang/zh-CN';
import enUS from '../../../src/components/locale/lang/en-US';

const langMap = {
  'en-US': {
    title: 'Xpre - Exploring Endless Possibilities',
    messages: enUS
  },
  'zh-CN': {
    title: 'Xpre | 一个差不多的组件库',
    messages: zhCN
  }
};

let currentLang = '';

function getDefaultLang() {
  const langs = Object.keys(langMap);
  const { hash } = location;

  for (let i = 0; i < langs.length; i++) {
    if (hash.indexOf(langs[i]) !== -1) {
      return langs[i];
    }
  }

  const userLang = localStorage.getItem('VANT_LANGUAGE') || navigator.language || 'en-US';
  return userLang.indexOf('zh-') !== -1 ? 'zh-CN' : 'en-US';
}

export function setLang(lang) {
  if (currentLang === lang) {
    return;
  }

  currentLang = lang;
  if (window.localStorage) {
    localStorage.setItem('VANT_LANGUAGE', lang);
  }
  Locale.use(lang, langMap[lang].messages);
  document.title = langMap[lang].title;
}

setLang(getDefaultLang());
