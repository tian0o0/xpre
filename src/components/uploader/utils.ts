export function toArray<T>(item: T | T[]): T[] {
  if (Array.isArray(item)) {
    return item;
  }

  return [item];
}

export function readFile(file: File, resultType: string) {
  return new Promise(resolve => {
    const reader = new FileReader();

    reader.onload = event => {
      resolve((event.target as FileReader).result);
    };

    if (resultType === 'dataUrl') {
      reader.readAsDataURL(file);
    } else if (resultType === 'text') {
      reader.readAsText(file);
    }
  });
}

export function isOversize(files: File | File[], maxSize: number): boolean {
  return toArray(files).some(file => file.size > maxSize);
}

export type FileListItem = {
  url?: string;
  file?: File;
  content?: string; // dataUrl
};

const IMAGE_EXT = ['jpeg', 'jpg', 'gif', 'png', 'svg'];

export function isImageUrl(url: string): boolean {
  return IMAGE_EXT.some(ext => url.indexOf(`.${ext}`) !== -1);
}

export function isImageFile(item: FileListItem): boolean {
  if (item.file && item.file.type) {
    return item.file.type.indexOf('image') === 0;
  }

  if (item.url) {
    return isImageUrl(item.url);
  }

  if (item.content) {
    return item.content.indexOf('data:image') === 0;
  }

  return false;
}
