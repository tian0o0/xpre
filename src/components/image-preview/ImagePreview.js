import { createNamespace } from '../../utils';
import { range } from '../../utils/format/number';
import { preventDefault } from '../../utils/dom/event';
import { PopupMixin } from '../../mixins/popup';
import { TouchMixin } from '../../mixins/touch';
import { CloseOnPopstateMixin } from '../../mixins/close-on-popstate';
import Slide from '../slide';

const [createComponent, bem] = createNamespace('image-preview');

function getDistance(touches) {
  return Math.sqrt(
    Math.abs(
      (touches[0].clientX - touches[1].clientX) *
        (touches[0].clientY - touches[1].clientY)
    )
  );
}

export default createComponent({
  mixins: [
    PopupMixin,
    TouchMixin,
    CloseOnPopstateMixin
  ],

  props: {
    className: null,
    lazyLoad: Boolean,
    asyncClose: Boolean,
    showIndicators: Boolean,
    images: {
      type: Array,
      default: () => []
    },
    loop: {
      type: Boolean,
      default: true
    },
    overlay: {
      type: Boolean,
      default: true
    },
    showIndex: {
      type: Boolean,
      default: true
    },
    startPosition: {
      type: Number,
      default: 0
    },
    minZoom: {
      type: Number,
      default: 1 / 3
    },
    maxZoom: {
      type: Number,
      default: 3
    },
    overlayClass: {
      type: String,
      default: bem('overlay')
    },
    closeOnClickOverlay: {
      type: Boolean,
      default: true
    }
  },

  data() {
    this.bindStatus = false;

    return {
      scale: 1,
      moveX: 0,
      moveY: 0,
      moving: false,
      zooming: false,
      active: 0,
      doubleClickTimer: null
    };
  },

  computed: {
    imageStyle() {
      const { scale } = this;
      const style = {
        transition: this.zooming || this.moving ? '' : '.3s all'
      };

      if (scale !== 1) {
        style.transform = `scale3d(${scale}, ${scale}, 1) translate(${this.moveX /
          scale}px, ${this.moveY / scale}px)`;
      }

      return style;
    }
  },

  watch: {
    value() {
      this.active = this.startPosition;
    },

    startPosition(active) {
      this.active = active;
    }
  },

  methods: {
    onWrapperTouchStart() {
      this.touchStartTime = new Date();
    },

    onWrapperTouchEnd(event) {
      preventDefault(event);

      const deltaTime = new Date() - this.touchStartTime;
      const { offsetX = 0, offsetY = 0 } = this.$refs.swipe || {};
      const GAP_TIME = 300;

      // prevent long tap to close component
      if (deltaTime < GAP_TIME && offsetX < 10 && offsetY < 10) {
        if (!this.doubleClickTimer) {
          this.doubleClickTimer = setTimeout(() => {
            const index = this.active;

            this.resetScale();
            this.$emit('close', {
              index,
              url: this.images[index]
            });

            if (!this.asyncClose) {
              this.$emit('input', false);
            }

            this.doubleClickTimer = null;
          }, GAP_TIME);
        } else {
          clearTimeout(this.doubleClickTimer);
          this.doubleClickTimer = null;
          this.toggleScale();
        }
      }
    },

    startMove(event) {
      const image = event.currentTarget;
      const rect = image.getBoundingClientRect();
      const winWidth = window.innerWidth;
      const winHeight = window.innerHeight;

      this.touchStart(event);
      this.moving = true;
      this.startMoveX = this.moveX;
      this.startMoveY = this.moveY;
      this.maxMoveX = Math.max(0, (rect.width - winWidth) / 2);
      this.maxMoveY = Math.max(0, (rect.height - winHeight) / 2);
    },

    startZoom(event) {
      this.moving = false;
      this.zooming = true;
      this.startScale = this.scale;
      this.startDistance = getDistance(event.touches);
    },

    onImageTouchStart(event) {
      const { touches } = event;
      const { offsetX = 0 } = this.$refs.swipe || {};

      if (touches.length === 1 && this.scale !== 1) {
        this.startMove(event);
      } /* istanbul ignore else */ else if (touches.length === 2 && !offsetX) {
        this.startZoom(event);
      }
    },

    onImageTouchMove(event) {
      const { touches } = event;
      if (this.moving || this.zooming) {
        preventDefault(event, true);
      }

      if (this.moving) {
        this.touchMove(event);
        const moveX = this.deltaX + this.startMoveX;
        const moveY = this.deltaY + this.startMoveY;
        this.moveX = range(moveX, -this.maxMoveX, this.maxMoveX);
        this.moveY = range(moveY, -this.maxMoveY, this.maxMoveY);
      }

      if (this.zooming && touches.length === 2) {
        const distance = getDistance(touches);
        const scale = (this.startScale * distance) / this.startDistance;
        this.scale = range(scale, this.minZoom, this.maxZoom);
      }
    },

    onImageTouchEnd(event) {
      /* istanbul ignore else */
      if (this.moving || this.zooming) {
        let stopPropagation = true;

        if (
          this.moving &&
          this.startMoveX === this.moveX &&
          this.startMoveY === this.moveY
        ) {
          stopPropagation = false;
        }

        if (!event.touches.length) {
          this.moving = false;
          this.zooming = false;
          this.startMoveX = 0;
          this.startMoveY = 0;
          this.startScale = 1;

          if (this.scale < 1) {
            this.resetScale();
          }
        }

        if (stopPropagation) {
          preventDefault(event, true);
        }
      }
    },

    onSwipeChange(active) {
      this.resetScale();
      this.active = active;
      this.$emit('change', active);
    },

    resetScale() {
      this.scale = 1;
      this.moveX = 0;
      this.moveY = 0;
    },

    toggleScale() {
      const scale = this.scale > 1 ? 1 : 2;

      this.scale = scale;
      this.moveX = 0;
      this.moveY = 0;
    }
  },

  render(h) {
    if (!this.value) {
      return;
    }

    const { active, images } = this;

    // const Index = this.showIndex && (
    //   <div class={bem('index')}>
    //     {this.slots('index') || `${active + 1}/${images.length}`}
    //   </div>
    // );

    const Images = (
      <Slide
        ref="swipe"
        loop={this.loop}
        autoplay={false}
      >
        {images.map((image, index) => {
          const props = {
            class: bem('image'),
            style: index === active ? this.imageStyle : null,
            // on: {
            //   touchstart: this.onImageTouchStart,
            //   touchmove: this.onImageTouchMove,
            //   touchend: this.onImageTouchEnd,
            //   touchcancel: this.onImageTouchEnd
            // }
          };
          return (
            this.lazyLoad ? (
                <img vLazy={image} {...props} />
            ) : (
                <img src={image} {...props} />
            )
          );
        })}
      </Slide>
    );

    return (
      <transition name="x-fade">
        <div
          class={[bem(), this.className]}
          onTouchstart={this.onWrapperTouchStart}
          onTouchMove={preventDefault}
          onTouchend={this.onWrapperTouchEnd}
          onTouchcancel={this.onWrapperTouchEnd}
        >
          {Images}
        </div>
      </transition>
    );
  }
});
