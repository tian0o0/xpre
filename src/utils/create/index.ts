import { createBEM, Bem } from './bem';
import { createComponent, Component } from './component';
import { createI18N, Translate } from './i18n';

type CreateNamespaceReturn = [
  Component,
  Bem,
  Translate
];

export function createNamespace(name: string): CreateNamespaceReturn {
  name = 'x-' + name;
  return [createComponent(name), createBEM(name), createI18N(name)];
}
