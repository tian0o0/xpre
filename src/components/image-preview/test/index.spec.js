import Vue from 'vue';
import ImagePreview from '..';
import ImagePreviewVue from '../ImagePreview';
import { mount, trigger, triggerDrag, transitionStub, later } from '../../../../test/utils';

transitionStub();

function triggerZoom(el, x, y) {
  trigger(el, 'touchstart', 0, 0, { x, y });
  trigger(el, 'touchmove', -x / 4, -y / 4, { x, y });
  trigger(el, 'touchmove', -x / 3, -y / 3, { x, y });
  trigger(el, 'touchmove', -x / 2, -y / 2, { x, y });
  trigger(el, 'touchmove', -x, -y, { x, y });
  trigger(el, 'touchend', 0, 0, { touchList: [] });
}

const images = [
  'https://img.yzcdn.cn/1.png',
  'https://img.yzcdn.cn/2.png',
  'https://img.yzcdn.cn/3.png'
];

test('render image', async () => {
  const wrapper = mount(ImagePreviewVue, {
    propsData: { images, value: true }
  });

  expect(wrapper).toMatchSnapshot();

  const swipe = wrapper.find('.van-swipe__track');
  triggerDrag(swipe, 500, 0);
  expect(wrapper.emitted('input')).toBeFalsy();
  triggerDrag(swipe, 0, 0);

  await later(300);

  expect(wrapper.emitted('input')[0][0]).toBeFalsy();
  expect(wrapper.emitted('change')[0][0]).toEqual(2);
});

test('async close', () => {
  const wrapper = mount(ImagePreviewVue, {
    propsData: {
      images,
      value: true,
      asyncClose: true
    }
  });

  const swipe = wrapper.find('.van-swipe__track');
  triggerDrag(swipe, 0, 0);
  expect(wrapper.emitted('input')).toBeFalsy();
  wrapper.vm.close();
  expect(wrapper.emitted('input')[0][0]).toBeFalsy();
});

test('function call', done => {
  ImagePreview(images);
  ImagePreview(images.slice(0, 1));
  Vue.nextTick(() => {
    const wrapper = document.querySelector('.van-image-preview');
    const swipe = wrapper.querySelector('.van-swipe__track');
    triggerDrag(swipe, 0, 0);

    expect(wrapper.querySelectorAll('img').length).toEqual(1);
    done();
  });
});

test('double click', async done => {
  const instance = ImagePreview(images);

  const swipe = instance.$el.querySelector('.van-swipe__track');
  triggerDrag(swipe, 0, 0);
  triggerDrag(swipe, 0, 0);
  expect(instance.scale).toEqual(2);

  await later();

  triggerDrag(swipe, 0, 0);
  triggerDrag(swipe, 0, 0);
  expect(instance.scale).toEqual(1);
  done();
});

test('onClose option', async done => {
  const onClose = jest.fn();
  const instance = ImagePreview({
    images,
    startPostion: 1,
    onClose
  });

  instance.$emit('input', true);
  expect(onClose).toHaveBeenCalledTimes(0);

  await later(300);

  const wrapper = document.querySelector('.van-image-preview');
  const swipe = wrapper.querySelector('.van-swipe__track');
  triggerDrag(swipe, 0, 0);
  expect(onClose).toHaveBeenCalledTimes(1);
  expect(onClose).toHaveBeenCalledWith({ index: 0, url: 'https://img.yzcdn.cn/1.png' });
  done();
});

test('onChange option', async done => {
  const instance = ImagePreview({
    images,
    startPostion: 0,
    onChange(index) {
      expect(index).toEqual(2);
      done();
    }
  });

  const swipe = instance.$el.querySelector('.van-swipe__track');
  triggerDrag(swipe, 1000, 0);
});

test('register component', () => {
  Vue.use(ImagePreview);
  expect(Vue.component(ImagePreviewVue.name)).toBeTruthy();
});

test('zoom', async () => {
  const { getBoundingClientRect } = Element.prototype;
  Element.prototype.getBoundingClientRect = jest.fn(() => ({ width: 100 }));

  const wrapper = mount(ImagePreviewVue, {
    propsData: { images, value: true }
  });

  const image = wrapper.find('img');
  triggerZoom(image, 300, 300);
  triggerDrag(image, 300, 300);
  expect(wrapper).toMatchSnapshot();
  Element.prototype.getBoundingClientRect = getBoundingClientRect;
});

test('index slot', () => {
  const wrapper = mount({
    template: `
      <van-image-preview :value="true">
        <template v-slot:index>Custom Index</template>
      </van-image-preview>
    `
  });

  expect(wrapper).toMatchSnapshot();
});

test('closeOnPopstate', () => {
  const wrapper = mount(ImagePreviewVue, {
    propsData: {
      images,
      value: true,
      closeOnPopstate: true
    }
  });

  trigger(window, 'popstate');
  expect(wrapper.emitted('input')[0][0]).toBeFalsy();

  wrapper.setProps({
    value: true,
    closeOnPopstate: false
  });

  trigger(window, 'popstate');
  expect(wrapper.emitted('input')[1]).toBeFalsy();
});

test('lazy-load prop', () => {
  const wrapper = mount(ImagePreviewVue, {
    propsData: {
      images,
      lazyLoad: true
    }
  });

  wrapper.setProps({
    value: true
  });

  expect(wrapper).toMatchSnapshot();
});
