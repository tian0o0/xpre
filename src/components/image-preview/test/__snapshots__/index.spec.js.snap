// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index slot 1`] = `<van-image-preview value="true"></van-image-preview>`;

exports[`lazy-load prop 1`] = `
<div class="x-image-preview" name="van-fade">
  <div class="x-image-preview__index">1/3</div>
  <div class="x-swipe">
    <div class="x-swipe__track" style="width: 0px; transition-duration: 0ms; transform: translateX(0px);">
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img class="x-image-preview__image" style="transition: .3s all;"></div>
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img class="x-image-preview__image"></div>
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img class="x-image-preview__image"></div>
    </div>
  </div>
</div>
`;

exports[`render image 1`] = `
<div class="x-image-preview" name="van-fade">
  <div class="x-image-preview__index">1/3</div>
  <div class="x-swipe">
    <div class="x-swipe__track" style="width: 0px; transition-duration: 500ms; transform: translateX(0px);">
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/1.png" class="x-image-preview__image" style="transition: .3s all;"></div>
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/2.png" class="x-image-preview__image"></div>
      <div class="x-swipe-item" style="width: 0px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/3.png" class="x-image-preview__image"></div>
    </div>
  </div>
</div>
`;

exports[`zoom 1`] = `
<div class="x-image-preview" name="van-fade">
  <div class="x-image-preview__index">1/3</div>
  <div class="x-swipe">
    <div class="x-swipe__track" style="transition-duration: 500ms; width: 0px; transform: translateX(0px);">
      <div class="x-swipe-item" style="width: 100px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/1.png" class="x-image-preview__image" style="transform: scale3d(2, 2, 1) translate(0px, NaNpx);"></div>
      <div class="x-swipe-item" style="width: 100px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/2.png" class="x-image-preview__image"></div>
      <div class="x-swipe-item" style="width: 100px; height: 100%; transform: translateX(0px);"><img src="https://img.yzcdn.cn/3.png" class="x-image-preview__image"></div>
    </div>
  </div>
</div>
`;
